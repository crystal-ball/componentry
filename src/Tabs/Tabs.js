import { useTheme } from '../Theme/Theme'
import { activeContainerComponent } from '../factories/active-container-component'
import { activeContentComponent } from '../factories/active-content-component'
import { activeTriggerComponent } from '../factories/active-trigger-component'
import { staticComponent } from '../factories/static-component'
import { element } from '../factories/element'
import { navClasses } from '../utils/componentry'

/**
 * [Tabs component 📝](https://componentry.design/components/tabs)
 */
export const Tabs = activeContainerComponent('tabs', { name: 'tabs' })

/**
 * [Tabs content container component 📝](https://componentry.design/components/tabs)
 */
Tabs.ContentContainer = staticComponent('TabsContentContainer', {
  componentCx: 'tabs-content-container',
})

/**
 * [Tabs content component 📝](https://componentry.design/components/tabs)
 */
Tabs.Content = activeContentComponent('tabs', {
  arias: { hidden: true, role: 'tabpanel' },
  variant: 'tabs-content',
})

/**
 * [Tabs nav component 📝](https://componentry.design/components/tabs)
 */
Tabs.TriggersContainer = function TriggersContainer(props) {
  return element({
    role: 'tablist',
    componentCx: navClasses('tabs-triggers-container', props),
    ...useTheme('TriggersContainer'),
    ...props,
  })
}
Tabs.TriggersContainer.displayName = 'TriggersContainer'

/**
 * [Tabs trigger component 📝](https://componentry.design/components/tabs)
 */
Tabs.Trigger = activeTriggerComponent('tabs', {
  arias: { selected: true, role: 'tab' },
  // Tabs can only activate, they never deactivate when clicked
  action: 'activate',
})
