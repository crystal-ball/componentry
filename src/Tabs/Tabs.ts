import { useTheme } from '../Theme/Theme'
import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionBuilder } from '../utils/active-action-component-builder'
import { activeContentBuilder } from '../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../utils/types'
import { navClasses } from '../utils/class-names'
import { element } from '../utils/element-creator'
import { staticComponent } from '../utils/static-component-builder'

export interface TabsProps extends ActiveContainerBaseProps, ComponentBaseProps<'div'> {}

export interface TabsActionsContainerProps extends ComponentBaseProps<'div'> {
  fill?: boolean
  justify?: boolean
  pills?: boolean
  vertical?: boolean
}

export interface TabsActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TabsContentContainerProps extends ComponentBaseProps<'div'> {}

export interface TabsContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tabs {
  (props: TabsProps): React.ReactElement
  /**
   * [Tabs actions container component üìù](https://componentry.design/components/tabs)
   */
  ActionsContainer: React.FC<TabsActionsContainerProps>
  /**
   * [Tabs action component üìù](https://componentry.design/components/tabs)
   */
  Action: React.FC<TabsActionProps>
  /**
   * [Tabs content container component üìù](https://componentry.design/components/tabs)
   */
  ContentContainer: React.FC<TabsContentContainerProps>
  /**
   * [Tabs content component üìù](https://componentry.design/components/tabs)
   */
  Content: React.FC<TabsContentProps>
}

/**
 * [Tabs component üìù](https://componentry.design/components/tabs)
 * @experimental
 */
export const Tabs = activeContainerBuilder<TabsProps>('Tabs') as Tabs

const ActionsContainer: React.FC<TabsActionsContainerProps> = (props) => {
  return element('TabsActionsContainer', {
    role: 'tablist',
    componentCx: navClasses('tabs-actions-container', props),
    ...useTheme<TabsActionsContainerProps>('TabsActionsContainer'),
    ...props,
  })
}
ActionsContainer.displayName = 'TabsActionsContainer'
Tabs.ActionsContainer = ActionsContainer

Tabs.Action = activeActionBuilder<TabsActionProps>('TabsAction', {
  aria: { selected: true, role: 'tab' },
  // Tabs can only activate, they never deactivate when clicked
  action: 'activate',
})

Tabs.ContentContainer = staticComponent<TabsContentContainerProps>('TabsContentContainer')

Tabs.Content = activeContentBuilder<TabsContentProps>('TabsContent', {
  aria: { hidden: true, role: 'tabpanel' },
})
