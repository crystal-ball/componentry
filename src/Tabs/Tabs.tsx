import { useTheme } from '../Theme/Theme'
import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionComponent } from '../utils/active-action-component-builder'
import { activeContentComponent } from '../utils/active-content-component-builder'
import {
  BaseActiveActionProps,
  BaseActiveContainerProps,
  BaseActiveContentProps,
  BaseProps,
} from '../utils/base-types'
import { navClasses } from '../utils/componentry'
import { element } from '../utils/element-creator'
import { staticComponent } from '../utils/static-component-builder'

export interface TabsProps
  extends BaseActiveContainerProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface TabsActionsContainerProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  fill?: boolean
  justify?: boolean
  pills?: boolean
  vertical?: boolean
}

export interface TabsActionProps
  extends BaseActiveActionProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TabsContentContainerProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface TabsContentProps
  extends BaseActiveContentProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tabs {
  (props: TabsProps): React.ReactElement
  /**
   * [Tabs actions container component üìù](https://componentry.design/components/tabs)
   */
  ActionsContainer: React.FC<TabsActionsContainerProps>
  /**
   * [Tabs action component üìù](https://componentry.design/components/tabs)
   */
  Action: React.FC<TabsActionProps>
  /**
   * [Tabs content container component üìù](https://componentry.design/components/tabs)
   */
  ContentContainer: React.FC<TabsContentContainerProps>
  /**
   * [Tabs content component üìù](https://componentry.design/components/tabs)
   */
  Content: React.FC<TabsContentProps>
}

/**
 * [Tabs component üìù](https://componentry.design/components/tabs)
 */
export const Tabs = activeContainerBuilder<TabsProps>('Tabs') as Tabs

const ActionsContainer: React.FC<TabsActionsContainerProps> = (props) => {
  return element({
    role: 'tablist',
    componentCx: navClasses('tabs-actions-container', props),
    ...useTheme<TabsActionsContainerProps>('TabsActionsContainer'),
    ...props,
  })
}
ActionsContainer.displayName = 'TabsActionsContainer'
Tabs.ActionsContainer = ActionsContainer

Tabs.Action = activeActionComponent<TabsActionProps>('TabsAction', {
  aria: { selected: true, role: 'tab' },
  // Tabs can only activate, they never deactivate when clicked
  action: 'activate',
})

Tabs.ContentContainer = staticComponent<TabsContentContainerProps>(
  'TabsContentContainer',
  {
    componentCx: 'üÖ≤-tabs-content-container',
  },
)

Tabs.Content = activeContentComponent<TabsContentProps>('TabsContent', {
  aria: { hidden: true, role: 'tabpanel' },
})
