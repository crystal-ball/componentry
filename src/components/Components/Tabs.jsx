import React from 'react'

import { Tab } from '../../../lib'

export default () => (
  <div className="mb-5">
    <div className="row">
      <div className="col-12">
        <p className="lead">Nav component...</p>
        <Tab defaultTab="three">
          <Tab.Nav className="nav-justified mb-3">
            <Tab.Trigger tabId="one">Item 1</Tab.Trigger>
            <Tab.Trigger tabId="two">Item 2</Tab.Trigger>
            <Tab.Trigger tabId="three">Item 3</Tab.Trigger>
          </Tab.Nav>
          <Tab.ContentContainer>
            <Tab.Content tabId="one">Tab 1</Tab.Content>
            <Tab.Content tabId="two">Tab 2</Tab.Content>
            <Tab.Content tabId="three">Tab 3</Tab.Content>
          </Tab.ContentContainer>
        </Tab>
      </div>
    </div>
  </div>
)
