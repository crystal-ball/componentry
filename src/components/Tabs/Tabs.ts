import { useTheme } from '../Theme/Theme'
import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  type ActiveActionBaseProps,
  type ActiveContainerBaseProps,
  type ActiveContentBaseProps,
  type ComponentBaseProps,
} from '../../utils/base-types'
import { element } from '../../utils/element-creator'
import { staticComponent } from '../../utils/static-component-builder'
import { Link } from '../Link/Link'

export interface TabsProps extends ActiveContainerBaseProps, ComponentBaseProps<'div'> {}

export interface TabsActionsContainerProps extends ComponentBaseProps<'div'> {
  /** Pill style tabs */
  pills?: boolean // TODO: This should be a variant on Tabs
}

export interface TabsActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TabsContentContainerProps extends ComponentBaseProps<'div'> {}

export interface TabsContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tabs {
  (props: TabsProps): React.ReactElement
  /**
   * [Tabs actions container component üìù](https://componentry.design/components/tabs)
   */
  ActionsContainer: React.FC<TabsActionsContainerProps>
  /**
   * [Tabs action component üìù](https://componentry.design/components/tabs)
   */
  Action: React.FC<TabsActionProps>
  /**
   * [Tabs content container component üìù](https://componentry.design/components/tabs)
   */
  ContentContainer: React.FC<TabsContentContainerProps>
  /**
   * [Tabs content component üìù](https://componentry.design/components/tabs)
   */
  Content: React.FC<TabsContentProps>
}

/**
 * [Tabs component üìù](https://componentry.design/components/tabs)
 * @experimental
 */
export const Tabs = activeContainerBuilder<TabsProps>('Tabs') as Tabs

const ActionsContainer: React.FC<TabsActionsContainerProps> = (props) => {
  const { pills, ...rest } = {
    ...useTheme<TabsActionsContainerProps>('TabsActionsContainer'),
    ...props,
  }

  return element({
    role: 'tablist',
    componentCx: [
      'C9Y-TabsContainer',
      {
        'C9Y-TabsContainer-pills': pills,
      },
    ],
    ...rest,
  })
}
ActionsContainer.displayName = 'TabsActionsContainer'
Tabs.ActionsContainer = ActionsContainer

Tabs.Action = activeActionBuilder<TabsActionProps>('TabsAction', {
  aria: { selected: true, role: 'tab' },
  // Tabs can only activate, they never deactivate when clicked
  action: 'activate',
  defaultAs: Link,
})

Tabs.ContentContainer = staticComponent<TabsContentContainerProps>('TabsContentContainer')

Tabs.Content = activeContentBuilder<TabsContentProps>('TabsContent', {
  aria: { hidden: true, role: 'tabpanel' },
})
