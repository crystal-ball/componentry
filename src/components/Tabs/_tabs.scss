/* componentry/src/Tabs/tabs */

//
// Tabs
//

// --- Structure
// .C9Y-tabs
// .C9Y-tabs-actions-container
// .C9Y-tabs-action
// .C9Y-tabs-content-container
// .C9Y-tabs-content

//
// Tabs container
//

// Container handles: margin between tabs and panes, and layout of tabs
.C9Y-tabs-actions-container {
  display: flex;
  margin: $tabs-nav-container-margin;
  border-bottom: $tabs-nav-container-border;
  background: $tabs-nav-container-bg;
  box-shadow: $tabs-nav-container-shadow;

  // --- Filled
  &.tabs-actions-container-fill {
    .tabs-action-primary {
      flex: 1 1 auto;
      text-align: center;
    }
  }

  // --- Justified
  &.tabs-actions-container-justified {
    .tabs-action-primary {
      flex-basis: 0;
      flex-grow: 1;
      text-align: center;
    }
  }

  // TODO: tabs pills

  // --- Vertical
  &.tabs-actions-container-vertical {
    margin: $tabs-nav-vertical-margin;
    border-bottom: none;

    .tabs-action-primary {
      margin: $tabs-tab-vertical-margin;
      border-width: $tabs-tab-vertical-border-width;
      border-color: $tabs-tab-vertical-border-color;
      border-radius: $tabs-tab-vertical-border-radius;

      transition: $tabs-tab-vertical-transition;

      &:hover,
      &:focus {
        border-width: $tabs-tab-vertical-hover-border-width;
        border-color: $tabs-tab-vertical-hover-border-color;
      }

      &.active {
        border-width: $tabs-tab-vertical-active-border-width;
        border-color: $tabs-tab-vertical-active-border-color;
      }

      &.disabled {
        border-width: $tabs-tab-vertical-disabled-border-width;
        border-color: $tabs-tab-vertical-disabled-border-color;
      }
    }
  }
}

//
// Tabs
//

// NOTE: Tabs are assumed to be an anchor or button element for accessibility!
.tabs-action-primary {
  -webkit-appearance: none; // reset buttons
  white-space: nowrap;
  padding: $tabs-tab-padding-y $tabs-tab-padding-x;
  background: $tabs-tab-bg;
  color: $tabs-tab-color;
  cursor: pointer;
  font-size: $font-size-base;
  font-weight: $font-weight-normal;
  border-style: solid;
  border-color: $tabs-tab-border-color;
  border-width: $tabs-tab-border-width;
  border-radius: $tabs-tab-border-radius;

  transition: $tabs-tab-transition;

  &:hover,
  &:focus {
    color: $tabs-tab-hover-color;
    background: $tabs-tab-hover-bg;
    border-color: $tabs-tab-hover-border-color;
  }

  &.active {
    background: $tabs-tab-active-bg;
    border-color: $tabs-tab-active-border-color;
    color: $tabs-tab-active-color;
  }

  &.disabled {
    background: $tabs-tab-disabled-bg;
    border-color: $tabs-tab-disabled-border-color;
    color: $tabs-tab-disabled-color;
  }
}

//
// Tab content container
//

.C9Y-tabs-content-container {
  padding: $tabs-pane-container-padding;
}
