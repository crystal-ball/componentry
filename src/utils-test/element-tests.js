/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
