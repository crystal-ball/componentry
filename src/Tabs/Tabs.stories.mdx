import { Meta } from '@storybook/addon-docs/blocks'
import { boolean } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import Tabs from './Tabs'

<Meta title='Components|Tabs' component={Tabs} />

# Tabs

<Tabs defaultActive='one' className='w-100'>
  <Tabs.Nav fill={boolean('fill', false)} justify={boolean('justify', false)}>
    <Tabs.Trigger activeId='one'>Item 1</Tabs.Trigger>
    <Tabs.Trigger activeId='two'>Tab with long name</Tabs.Trigger>
    <Tabs.Trigger activeId='three'>Item 3</Tabs.Trigger>
    <Tabs.Trigger activeId='four' disabled>
      Disabled
    </Tabs.Trigger>
  </Tabs.Nav>
  <Tabs.ContentContainer>
    <Tabs.Content activeId='one'>Tab 1</Tabs.Content>
    <Tabs.Content activeId='two'>Tab 2</Tabs.Content>
    <Tabs.Content activeId='three'>Tab 3</Tabs.Content>
    <Tabs.Content activeId='four'>This tab has been disabled</Tabs.Content>
  </Tabs.ContentContainer>
</Tabs>
