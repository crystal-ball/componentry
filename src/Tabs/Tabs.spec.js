import React from 'react'
import { render } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { Tabs } from './Tabs'

describe('<Tab />', () => {
  elementTests(Tabs)
  elementTests(Tabs.TriggersContainer)
  elementTests(Tabs.Trigger)
  elementTests(Tabs.ContentContainer)
  elementTests(Tabs.Content)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Tab /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Tabs defaultActive='one'>
        <Tabs.TriggersContainer>
          <Tabs.Trigger activeId='one'>Item 1</Tabs.Trigger>
          <Tabs.Trigger activeId='two'>Tab with long name</Tabs.Trigger>
          <Tabs.Trigger activeId='three'>Item 3</Tabs.Trigger>
          <Tabs.Trigger activeId='four' disabled>
            Disabled
          </Tabs.Trigger>
        </Tabs.TriggersContainer>
        <Tabs.ContentContainer>
          <Tabs.Content activeId='one'>Tab 1</Tabs.Content>
          <Tabs.Content activeId='two'>Tab 2</Tabs.Content>
          <Tabs.Content activeId='three'>Tab 3</Tabs.Content>
          <Tabs.Content activeId='four'>This tab has been disabled</Tabs.Content>
        </Tabs.ContentContainer>
      </Tabs>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
