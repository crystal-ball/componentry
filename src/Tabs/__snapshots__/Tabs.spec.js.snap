// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> snapshots renders correctly 1`] = `
<div
  class="tabs"
  data-id="guid"
  name="tabs"
>
  <div
    class="tabs-triggers-container"
    role="tablist"
  >
    <button
      aria-controls="guid-one-content"
      aria-selected="true"
      class="tabs-trigger active"
      id="guid-one-tab"
      role="tab"
      type="button"
      value="one"
    >
      Item 1
    </button>
    <button
      aria-controls="guid-two-content"
      aria-selected="false"
      class="tabs-trigger"
      id="guid-two-tab"
      role="tab"
      type="button"
      value="two"
    >
      Tab with long name
    </button>
    <button
      aria-controls="guid-three-content"
      aria-selected="false"
      class="tabs-trigger"
      id="guid-three-tab"
      role="tab"
      type="button"
      value="three"
    >
      Item 3
    </button>
    <button
      aria-controls="guid-four-content"
      aria-selected="false"
      class="tabs-trigger disabled"
      disabled=""
      id="guid-four-tab"
      role="tab"
      type="button"
      value="four"
    >
      Disabled
    </button>
  </div>
  <div
    class="tabs-content-container"
  >
    <div
      aria-hidden="false"
      aria-labelledby="guid-one-trigger"
      class="tabs-content"
      id="guid-one-content"
      role="tabpanel"
    >
      Tab 1
    </div>
    <div
      aria-hidden="true"
      aria-labelledby="guid-two-trigger"
      class="tabs-content"
      id="guid-two-content"
      role="tabpanel"
    >
      Tab 2
    </div>
    <div
      aria-hidden="true"
      aria-labelledby="guid-three-trigger"
      class="tabs-content"
      id="guid-three-content"
      role="tabpanel"
    >
      Tab 3
    </div>
    <div
      aria-hidden="true"
      aria-labelledby="guid-four-trigger"
      class="tabs-content"
      id="guid-four-content"
      role="tabpanel"
    >
      This tab has been disabled
    </div>
  </div>
</div>
`;
